### ISSUE TITLE
RolesAllowed annotations on interface are ignored

### ISSUE THREAD

**Issue by hellcats88** [quarkusio/quarkus#16840]
## Describe the bug
The @RolesAllowed annotation on the interface method that describes the REST API is ignored.
The only way to enable the role check is to add the annotation on the class method of the API implementation.

### Expected behavior
Usually the description of the API (Swagger, input/output, queryparams and other API descriptions are placed on the interface method to make the implementation cleaner.

### Actual behavior
Every call to an api annotated with @RolesAllowed on the interface method is managed as @PermitAll behavior

## To Reproduce

[security-openid-connect-quickstart.zip](https://github.com/quarkusio/quarkus/files/6385484/security-openid-connect-quickstart.zip)

Link to a small reproducer (preferably a Maven project if the issue is not Gradle-specific).

Or attach an archive containing the reproducer to the issue.

1.  Follow quarkus quickstart https://github.com/quarkusio/quarkus-quickstarts/tree/main/security-openid-connect-quickstart using the attached zip project
2. See that /api/users/me is called without errors 

### Output of `uname -a` or `ver`
Linux 2019-150507 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.20.04, mixed mode)

### Quarkus version or git rev
1.13.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
OS name: "linux", version: "5.4.72-microsoft-standard-wsl2", arch: "amd64", family: "unix"

## Additional context
(Add any other context about the problem here.)

**Comment by sberyozkin** [comment]
Hi Michal @michalszynkiewicz, Stuart @stuartwdouglas, 
It rings a bell, I recall there was some discussion about not supporting RolesAllowed inheritance - do you recall if it was made intentionally ? I guess yes as otherwise one can have the endpoint implementing N interfaces, each of them bringing the conflicting RBAC rules, so I'd say this is an out of box expectation that these rules are set on the actual endpoint class

**Comment by sberyozkin** [comment]
Hi @hellcats88 I've updated the subject to make it more specific as it does not only apply to OIDC but to the authorization layer in general

**Comment by mickroll** [comment]
Found the issue you remembered from the past: #12736

**Comment by famod** [comment]
Btw, this is a really unfortunate limitation when generating interfaces from OpenAPI files (json/yaml) via e.g. https://github.com/OpenAPITools/openapi-generator

**Comment by stuartwdouglas** [comment]
@mkouba what do you think about allowing interceptor bindings on interface methods as a CDI extension?

**Comment by mkouba** [comment]
> @mkouba what do you think about allowing interceptor bindings on interface methods as a CDI extension?

I'm not sure TBH. The spec is very clear that it **should not** be supported. For example, the [Inheritance and specialization](https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#inheritance) section mentions: 
* _"Type-level metadata is never inherited from interfaces implemented by a bean."_, and
* _"Suppose a class X is extended directly or indirectly by the bean class of a managed bean Y. If X is annotated with an interceptor binding type Z then Y inherits the annotation if and only if Z declares the `@Inherited` meta-annotation and neither Y nor any intermediate class that is a subclass of X and a superclass of Y declares an annotation of type Z. (This behavior is defined by the Java Language Specification.)"_  

And it would definitely break some existing code.

CC @manovotn @Ladicek

**Comment by manovotn** [comment]
> I'm not sure TBH. The spec is very clear that it should not be supported. For example, the Inheritance and specialization section mentions:

+1, spec literally says this should not work.
It is not the first case where we are discussing this though. Enabling it globally presents fairly huge compatibility risk and having it in just some places makes it inconsistent which is even weirder (at least to me).

@mkouba don't we already have some special handling for something like REST? I am pretty sure we had to deal with this already.

> How would that even work if an interceptor binding annotation for one method is "inherited" from multiple interfaces? Would we use our existing support for repeatable interceptor bindings on methods?

Good point, this is likely one of the reasons why the limitation exists. You'd need to change this on the common annotations spec level first (not that I am for it :)).

**Comment by stuartwdouglas** [comment]
I think the main issue here is that JAX-RS annotations are inherited from interfaces, so it is problematic that the security ones are not. I think it probably makes the most amount of sense to special case this for JAX-RS?

**Comment by mkouba** [comment]
> I think the main issue here is that JAX-RS annotations are inherited from interfaces, so it is problematic that the security ones are not. I think it probably makes the most amount of sense to special case this for JAX-RS?

AFAIK the bean validation supports them as well but there's some workaround in the quarkus extension. CC @gsmet

**Comment by mkouba** [comment]
> @mkouba don't we already have some special handling for something like REST? I am pretty sure we had to deal with this already.

Interceptor bindings are supported on MP Rest Client interfaces but that's a bit different because there's no bean class implementation.

**Comment by michalszynkiewicz** [comment]
One problem with adding support for inheriting security annotations from interfaces for jax-rs endpoints is that users most likely will assume that it will also work for usual CDI beans.
We could add some annotation, e.g. `@InheritedSecurity` that could be added to any bean and would mean that things like `@RolesAllowed` are inherited. But I'm not sure if it won't increase the mess :)

**Comment by mickroll** [comment]
With this info from the spec, a `@InheritedSecurity` annotation sounds like the way to go.
Or a more generic annotation like `@Inherit({RolesAllowed.class, Authenticated.class, PermitAll.class, MyCustomAnnotation.class})` would come in handy for the spec-first api approach. With `@Inherit` working like having `@Inherited` on the annotation itself.

**Comment by MartinX3** [comment]
I think I can confirm it with 2.5.1.
My class has `@RolesAllowed(xxx)`
My function has `@PermitAll`
And the `@PermitAll` is ignored. It always wants my user to be in a role.

**Comment by sberyozkin** [comment]
Let me close this issue as the discussion has moved to #22540

**Comment by wesleyeduardodev** [comment]
I went through this same problem and found it via debug. What happens is that the @RolesAllowed annotation on the interface is not passed on to the implementation. Therefore, the authentication flow uses the PermitAllCheck class from the quarkus-security package when it should use RolesAllowedCheck.

I forwarded the annotation to the method that actually has the interface implementation and with that the flow went through the RolesAllowedCheck class. At that moment I discovered another failure that I don't know if anyone has already experienced, which would be the case where the token's roles are not validated. That is, if I have a token with the role uma_authorization and I have an endpoint with the annotation @RolesAllowed("uma_authorization"), the return should be 200 ok but I was only getting a 403. This happened because my SecurityIdentity did not load the roles of the token, were always empty so the validation returned disallowed (403).

To fix it I used a quarkus interceptor class where at that moment I could extract the roles from the token and add them to the SecurityIdentity so that the RolesAllowed annotation actually works


@ApplicationScoped
public class RolesAugmentor implements SecurityIdentityAugmentor {

    @Override
    public Uni<SecurityIdentity> augment(SecurityIdentity identity, AuthenticationRequestContext context) {
        return Uni.createFrom().item(build(identity));
    }

    private Supplier<SecurityIdentity> build(SecurityIdentity identity) {
        if (identity.isAnonymous()) {
            return () -> identity;
        } else {
            QuarkusSecurityIdentity.Builder builder = QuarkusSecurityIdentity.builder(identity);

            //TODO Custom code to extract roles from token
            List<String> rolesDoToken = RoleUtils.returnTokenRoles(identity);

            rolesDoToken.forEach(builder::addRole);
            return builder::build;
        }
    }
}
